File: TriviaClient.py
import socket
import json

SERVER_PORT = 8826
BYTES_SIZE = 1024

ERROR_CODE = 0
SUCCESS_CODE = 1

LOGIN_CODE = 1
SIGNUP_CODE = 2
LOGOUT_CODE = 3
ROOMS_LIST_CODE = 4
PLAYERS_IN_ROOM_CODE = 5
HIGH_SCORES_CODE = 6
PERSONAL_STATS_CODE = 7
JOIN_ROOM_CODE = 8
CREATE_ROOM_CODE = 9
CLOSE_ROOM_CODE = 10
START_GAME_CODE = 11
ROOM_STATE_CODE = 12
LEAVE_ROOM_CODE = 13
LEAVE_GAME_CODE = 14
GET_QUESTION_CODE = 15
SUBMIT_ANSWER_CODE = 16
GET_GAME_RESULTS_CODE = 17
PLAYER_RESULTS_CODE = 18


def build_message(code: int, message: str) -> bytes:
    """
    Build a message to send to the server.
    :param code: 1 byte
    :param message: string
    :return: bytes
    """
    return (
            code.to_bytes(1, "big")
            + len(message).to_bytes(4, "big")
            + message.encode()
    )


def build_message_from_dict(code: int, message: dict) -> bytes:
    """
    Build a message to send to the server.
    :param code: 1 byte
    :param message: string
    :return: bytes
    """
    return build_message(code, json.dumps(message))


def parse_message(message: bytes) -> dict:
    """
    Parse a message from the server.
    :param message: bytes
    :return: dict
    """
    code = int.from_bytes(message[:1], "big")
    msg_len = int.from_bytes(message[1:5], "big")
    msg = message[5:].decode()
    return {"code": code, "len": msg_len, "msg": msg}


def parse_message_to_dict(message: bytes) -> dict:
    """
    Parse a message from the server.
    :param message: bytes
    :return: dict
    """
    parsed_message = parse_message(message)
    parsed_message["msg"] = json.loads(parsed_message["msg"])
    return parsed_message


class TriviaClient:
    """
    A class representing a Trivia client.
    """

    def __init__(self):
        self.client = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.client.connect(("localhost", SERVER_PORT))
        msg = self.client.recv(BYTES_SIZE)  # skip the welcome message

    def send_request_bytes(self, message: bytes) -> bytes:
        """
        Send a request to the server and return the response.
        :param message: message to send to the server
        :return: None
        """
        self.client.send(message)
        return self.client.recv(BYTES_SIZE)

    def send_request(self, code: int, message: str) -> dict:
        """
        Send a request to the server and return the response.
        :param code: 1 byte
        :param message: string
        :return: dict
        """
        return parse_message(self.send_request_bytes(build_message(code, message)))

    def send_request_dict(self, code: int, message: dict) -> dict:
        """
        Send a request to the server and return the response.
        :param code: 1 byte
        :param message: string
        :return: dict
        """
        return parse_message_to_dict(self.send_request_bytes(build_message_from_dict(code, message)))

    def login(self, username: str, password: str) -> dict:
        """
        Login to the server.
        :param username: username
        :param password: password
        :return: dict
        """
        data = {"username": username, "password": password}
        return self.send_request_dict(LOGIN_CODE, data)

    def signup(self, username: str, password: str, email: str) -> dict:
        """
        Signup to the server.
        :param username: username
        :param password: password
        :param email: email
        :return: dict
        """
        data = {"username": username, "password": password, "email": email}
        return self.send_request_dict(SIGNUP_CODE, data)

    def logout(self) -> dict:
        """
        Logout from the server.
        :return: dict
        """
        return self.send_request_dict(LOGOUT_CODE, {})

    def get_rooms_list(self) -> dict:
        """
        Get the rooms list.
        :return: dict
        """
        return self.send_request_dict(ROOMS_LIST_CODE, {})

    def get_players_in_room(self, room_id: int) -> dict:
        """
        Get the players in a room.
        :param room_id: room id
        :return: dict
        """
        return self.send_request_dict(PLAYERS_IN_ROOM_CODE, {"roomId": room_id})

    def get_high_scores(self) -> dict:
        """
        Get the high scores.
        :return: dict
        """
        return self.send_request_dict(HIGH_SCORES_CODE, {})

    def get_personal_stats(self) -> dict:
        """
        Get the personal stats.
        :return: dict
        """
        return self.send_request_dict(PERSONAL_STATS_CODE, {})

    def join_room(self, room_id: int) -> dict:
        """
        Join a room.
        :param room_id: room id
        :return: dict
        """
        return self.send_request_dict(JOIN_ROOM_CODE, {"roomId": room_id})

    def create_room(self, room_name: str, max_players: int, question_count: int, answer_timeout: int) -> dict:
        """
        Create a room.
        :param room_name: room name
        :param max_players: max players
        :param question_count: question count
        :param answer_timeout: answer timeout
        :return: dict
        """
        return self.send_request_dict(CREATE_ROOM_CODE, {
            "roomName": room_name,
            "maxUsers": max_players,
            "questionCount": question_count,
            "answerTimeout": answer_timeout
        })

    def leave_room(self) -> dict:
        """
        Leave a room.
        :return: dict
        """
        return self.send_request_dict(LEAVE_ROOM_CODE, {})

    def start_game(self) -> dict:
        """
        Start a game.
        :return: dict
        """
        return self.send_request_dict(START_GAME_CODE, {})

    def get_question(self) -> dict:
        """
        Get a question.
        :return: dict
        """
        return self.send_request_dict(GET_QUESTION_CODE, {})

    def submit_answer(self, answer_id: int) -> dict:
        """
        Submit an answer.
        :param answer_id: answer id
        :return: dict
        """
        return self.send_request_dict(SUBMIT_ANSWER_CODE, {"answerId": answer_id})

    def get_game_results(self) -> dict:
        """
        Get the game results.
        :return: dict
        """
        return self.send_request_dict(GET_GAME_RESULTS_CODE, {})

    def leave_game(self) -> dict:
        """
        Leave a game.
        :return: dict
        """
        return self.send_request_dict(LEAVE_GAME_CODE, {})


def test_signup(username: str, password: str, email: str):
    """
    Test login.
    :return: the client
    """
    client = TriviaClient()
    res = client.signup(username, password, email)
    if res["msg"]["status"] != SUCCESS_CODE:
        raise ValueError("Signup failed!")
    return client


def test_login(username: str, password: str):
    """
    Test login.
    :param username: username
    :param password: password
    :return: the client
    """
    client = TriviaClient()
    res = client.login(username, password)
    if res["code"] == ERROR_CODE:
        raise ValueError("Login failed!")
    return client


def prompt_login():
    """
    Prompt the user to login.
    :return: the client
    """
    username = input("Username: ")
    password = input("Password: ")
    return test_login(username, password)


def test():
    print("Please login to the server")
    me = prompt_login()
    print("Logged in successfully!")

    print("Testing stats...")

    print(me.get_personal_stats())
    print(me.get_high_scores())

    print("Testing rooms...")
    print(me.create_room("test", 2, 2, 2))
    print(me.get_rooms_list())

    print("Testing joining room... for this test to work, you need to login with another user and join room")
    me2 = prompt_login()
    print(me2.join_room(1))
    print(me2.get_players_in_room(1))

    print("Testing logout...")
    print(me.logout())
    print(me2.logout())


if __name__ == '__main__':
    me = prompt_login()
    me.create_room("test", 8, 8, 8)
    me.start_game()

File: TrivyaLidanMatan/TriviaClientApp/TriviaClient.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Net.Sockets;
using System.Text;
using System.Threading.Tasks;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace TriviaClientApp
{
    public class TriviaClient
    {
        public const string SERVER_IP = "localhost";
        public const int SERVER_PORT = 8826;
        public const int BYTE_SIZE = 1024;
        public const int TIMEOUT = 60000;

        public const int ERROR_CODE = 0;
        public const int SUCCESS_CODE = 1;

        public enum RequestCodes
        {
            ERROR_CODE = 0,
            LOGIN_CODE = 1,
            SIGNUP_CODE = 2,
            LOGOUT_CODE = 3,
            ROOMS_LIST_CODE = 4,
            PLAYERS_IN_ROOM_CODE = 5,
            HIGH_SCORES_CODE = 6,
            PERSONAL_STATS_CODE = 7,
            JOIN_ROOM_CODE = 8,
            CREATE_ROOM_CODE = 9,
            CLOSE_ROOM_CODE = 10,
            START_GAME_CODE = 11,
            ROOM_STATE_CODE = 12,
            LEAVE_ROOM_CODE = 13,
            LEAVE_GAME_CODE = 14,
            GET_QUESTION_CODE = 15,
            SUBMIT_ANSWER_CODE = 16,
            GET_GAME_RESULTS_CODE = 17,
            PLAYER_RESULTS_CODE = 18,
            ADD_QUESTION_CODE = 19,
            HTH_GET_STATE_CODE = 20,
            JOIN_HTH_CODE = 21,
            LEAVE_HTH_CODE = 22,
        }

        private static TriviaClient? instance = null;

        /// <summary>
        /// Get the Client
        /// If the client is not connected it will connect it
        /// </summary>
        /// <returns>The Client</returns>
        public static TriviaClient GetClient()
        {
            if (instance == null)
            {
                instance = new TriviaClient();
                instance.Connect();
            }

            return instance;
        }

        private Socket socket;

        public string Username { get; private set; }

        public TriviaClient()
        {
            socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
            socket.ReceiveTimeout = TIMEOUT;
            socket.SendTimeout = TIMEOUT;
            this.Username = "";
        }

        /// <summary>
        /// Build a message to send to the server.
        /// </summary>
        /// <param name="code">message code</param>
        /// <param name="message">message</param>
        /// <returns>byte list</returns>
        public List<byte> BuildMessage(byte code, string message)
        {
            List<byte> bytes = new List<byte>(5 + message.Length); // allocate memory on the heap
            bytes.Add(code);
            byte[] lengthBytes = BitConverter.GetBytes(message.Length);
            Array.Reverse(lengthBytes);
            bytes.AddRange(lengthBytes);
            bytes.AddRange(Encoding.ASCII.GetBytes(message));
            return bytes;
        }

        /// <summary>
        /// Build a message to send to the server.
        /// </summary>
        /// <param name="code">message code</param>
        /// <param name="dict">message dict</param>
        /// <returns>byte list</returns>
        public List<byte> BuildMessageFromDict(byte code, JObject dict)
        {
            string json = JsonConvert.SerializeObject(dict);
            return BuildMessage(code, json);
        }

        /// <summary>
        /// Parse a message from the server.
        /// </summary>
        /// <param name="bytes">message bytes</param>
        /// <returns>message dict</returns>
        public JObject ParseMessage(List<byte> bytes)
        {
            byte code = bytes[0];
            List<byte> lengthBytes = bytes.GetRange(1, 4);
            lengthBytes.Reverse();
            int length = BitConverter.ToInt32(lengthBytes.ToArray());
            List<byte> messageBytes = bytes.GetRange(5, length);
            string message = Encoding.ASCII.GetString(messageBytes.ToArray());
            JObject dict = new JObject
            {
                {"code", code},
                {"length", length},
                {"message", message}
            };
            return dict;
        }

        /// <summary>
        /// Parse a message from the server.
        /// </summary>
        /// <param name="bytes">message bytes</param>
        /// <returns>message dict</returns>
        public JObject ParseMessageToDict(List<byte> bytes)
        {
            var dict = ParseMessage(bytes);
            dict["message"] = JsonConvert.DeserializeObject<JObject>((string)dict["message"]);
            return dict;
        }

        /// <summary>
        /// Connect to the server.
        /// </summary>
        public void Connect()
        {
            try
            {
                socket.Connect(SERVER_IP, SERVER_PORT);
                Span<byte> buffer = new byte[BYTE_SIZE];
                socket.Receive(buffer);
            }
            catch (Exception e)
            {
                Debug.WriteLine(e);
                MessageBox.Show("Failed to connect to server: " + e.Message, "Connection ERROR", MessageBoxButtons.OK, MessageBoxIcon.Error);
                Environment.Exit(1);
            }
        }

        /// <summary>
        /// Send a request to the server and return the response.
        /// </summary>
        /// <param name="message">message to send to the server</param>
        /// <returns>response</returns>
        private List<byte> SendRequestBytes(List<byte> message)
        {
            try
            {
                socket.Send(message.ToArray());
            }
            catch (Exception e)
            {
                Debug.WriteLine(e);
                MessageBox.Show("Failed to send to server: " + e.Message, "Connection ERROR", MessageBoxButtons.OK, MessageBoxIcon.Error);
                Environment.Exit(1);
            }

            return ReceiveResponseBytes();
        }

        /// <summary>
        /// Receive Response Bytes
        /// </summary>
        /// <returns>response</returns>
        private List<byte> ReceiveResponseBytes()
        {
            byte[] buffer = new byte[BYTE_SIZE];
            List<byte> receivedData = new List<byte>();
            try
            {
                int bytesRead = 0;

                do
                {
                    bytesRead = socket.Receive(buffer);
                    receivedData.AddRange(buffer.Take(bytesRead));
                } while (bytesRead >= BYTE_SIZE);
            }
            catch (Exception e)
            {
                Debug.WriteLine(e);
                MessageBox.Show("Failed to receive from server: " + e.Message, "Connection ERROR", MessageBoxButtons.OK, MessageBoxIcon.Error);
                Environment.Exit(1);
            }
            return receivedData;
        }

        /// <summary>
        /// Send a request to the server and return the response.
        /// </summary>
        /// <param name="code">request code</param>
        /// <param name="message">the message</param>
        /// <returns></returns>
        private JObject SendRequest(byte code, string message = "")
        {
            return ParseMessage(SendRequestBytes(BuildMessage(code, message)));
        }

        /// <summary>
        /// Send a request to the server and return the response.
        /// </summary>
        /// <param name="code">request code</param>
        /// <param name="dict">request data</param>
        /// <returns>response</returns>
        public JObject SendRequestDict(byte code, JObject dict = null)
        {
            if (dict == null)
            {
                dict = new JObject();
            }
            return ParseMessageToDict(SendRequestBytes(BuildMessageFromDict(code, dict)));
        }

        /// <summary>
        /// Login to the server.
        /// {"status" : status}
        /// </summary>
        /// <param name="username">username</param>
        /// <param name="password">password</param>
        /// <returns>response</returns>
        public JObject Login(string username, string password)
        {
            JObject data = new JObject() { { "username", username }, { "password", password } };
            JObject result = SendRequestDict((byte)RequestCodes.LOGIN_CODE, data);
            if ((int) result["code"] != ERROR_CODE)
            {
                this.Username = username;
            }
            return result;
        }

        /// <summary>
        /// Signup to the server.
        /// {"status" : status}
        /// </summary>
        /// <param name="username">username</param>
        /// <param name="password">password</param>
        /// <param name="email">email</param>
        /// <returns>response</returns>
        public JObject Signup(string username, string password, string email)
        {
            JObject data = new JObject() { { "username", username }, { "password", password }, { "email", email } };
            JObject result = SendRequestDict((byte)RequestCodes.SIGNUP_CODE, data);
            if ((int)result["code"] != ERROR_CODE)
            {
                this.Username = username;
            }
            return result;
        }

        /// <summary>
        /// Logout from the server.
        /// {"status" : status}
        /// </summary>
        /// <returns>response</returns>
        public JObject Logout()
        {
            Username = "";
            return SendRequestDict((byte)RequestCodes.LOGOUT_CODE);
        }

        /// <summary>
        /// Get the room list.
        /// {"status" : status, "rooms" : {"isActive" : isActive, "name" : name, "maxPlayers" : maxPlayers, "numOfQuestionsInGame" : numOfQuestionsInGame, "timePerQuestion" : timePerQuestion, "id" : id}}
        /// </summary>
        /// <returns>response</returns>
        public JObject GetRoomsList()
        {
            return SendRequestDict((byte)RequestCodes.ROOMS_LIST_CODE);
        }

        /// <summary>
        /// Get the players in a room.
        /// {"status" : status, "players" : players}
        /// </summary>
        /// <param name="roomId">room id</param>
        /// <returns>response</returns>
        public JObject GetPlayersInRoom(int roomId)
        {
            JObject data = new JObject();
            data["roomId"] = roomId;
            return SendRequestDict((byte)RequestCodes.PLAYERS_IN_ROOM_CODE, data);
        }

        /// <summary>
        /// Get the high scores.
        /// {"status" : status, "highscores" : players}
        /// </summary>
        /// <returns>response</returns>
        public JObject GetHighScores()
        {
            return SendRequestDict((byte)RequestCodes.HIGH_SCORES_CODE);
        }

        /// <summary>
        /// Get the personal stats.
        /// {status" : status, "statistics" : statistics}
        /// </summary>
        /// <returns>response</returns>
        public JObject GetPersonalStats()
        {
            return SendRequestDict((byte)RequestCodes.PERSONAL_STATS_CODE);
        }

        /// <summary>
        /// Join a room.
        /// {"status" : status}
        /// </summary>
        /// <param name="roomId">room id</param>
        /// <returns>response</returns>
        public JObject JoinRoom(int roomId)
        {
            JObject data = new JObject() { { "roomId", roomId } };
            return SendRequestDict((byte)RequestCodes.JOIN_ROOM_CODE, data);
        }

        /// <summary>
        /// Create a room.
        /// {"status" : status}
        /// </summary>
        /// <param name="roomName">room name</param>
        /// <param name="maxUsers">max players</param>
        /// <param name="questionCount">question count</param>
        /// <param name="answerTimeout">answer timeout</param>
        /// <returns>response</returns>
        public JObject CreateRoom(string roomName, int maxUsers, int questionCount, int answerTimeout)
        {
            JObject data = new JObject() { { "roomName", roomName }, { "maxUsers", maxUsers }, { "questionCount", questionCount }, { "answerTimeout", answerTimeout } };
            return SendRequestDict((byte)RequestCodes.CREATE_ROOM_CODE, data);
        }

        /// <summary>
        /// Will get room state
        ///  {"status" : status, "has
        /// Begun" : hasGameBegun, "players" : players, "questionCount" : questionCount, "answerTimeout" : answerTimeout}
        /// </summary>
        /// <returns>response</returns>
        public JObject GetRoomState()
        {
            return SendRequestDict((byte)RequestCodes.ROOM_STATE_CODE);
        }

        /// <summary>
        /// Will leave the room
        /// {"status" : status}
        /// </summary>
        /// <returns>response</returns>
        public JObject LeaveRoom(){
            return SendRequestDict((byte)RequestCodes.LEAVE_ROOM_CODE);
        }

        /// <summary>
        /// Will close the room
        /// {"status" : status}
        /// </summary>
        /// <returns>response</returns>
        public JObject CloseRoom()
        {
            return SendRequestDict((byte)RequestCodes.CLOSE_ROOM_CODE);
        }

        /// <summary>
        /// Will start the game
        /// {"status" : status}
        /// </summary>
        /// <returns>response</returns>
        public JObject StartGame()
        {
            return SendRequestDict((byte)RequestCodes.START_GAME_CODE);
        }

        /// <summary>
        /// Will get the question
        /// {"status" : status, "question" : question, "answers" : [["answerNumber", "Answer"], ["answerNumber", "Answer"], ["answerNumber", "Answer"], ["answerNumber", "Answer"]]}
        /// </summary>
        /// <returns>response</returns>
        public JObject GetQuestion()
        {
            return SendRequestDict((byte)RequestCodes.GET_QUESTION_CODE);
        }

        /// <summary>
        /// Will submit the answer
        /// {"status" : status, "correctAnswer" : correctAnswer}
        /// </summary>
        /// <param name="answerId">answer id</param>
        /// <returns>response</returns>
        public JObject SubmitAnswer(int answerId)
        {
            JObject data = new JObject() { { "answerId", answerId } };
            return SendRequestDict((byte)RequestCodes.SUBMIT_ANSWER_CODE, data);
        }

        /// <summary>
        /// Will get the game results
        /// {"status" : status, "results" :
        /// {"username" : username, "correctAnswerCount" : correctAnswerCount", "wrongAnswerCount" : wrongAnswerCount, "AverageAnswerTime" : AverageAnswerTime}}
        /// </summary>
        /// <returns>response</returns>
        public JObject GetGameResults()
        {
            return SendRequestDict((byte)RequestCodes.GET_GAME_RESULTS_CODE);
        }

        /// <summary>
        /// Will leave the game
        /// {"status" : status}
        /// </summary>
        /// <returns>response</returns>
        public JObject LeaveGame()
        {
            return SendRequestDict((byte)RequestCodes.LEAVE_GAME_CODE);
        }

        /// <summary>
        /// Will get the room state
        /// </summary>
        /// <returns>response</returns>
        public JObject GetRoomUsers()
        {
            return SendRequestDict((byte)RequestCodes.PLAYER_RESULTS_CODE);
        }

        /// <summary>
        /// Will add a question
        /// {"status" : status}
        /// </summary>
        /// <returns>response</returns>
        public JObject AddQuestion(string question, string correctAnswer, string answer2, string answer3, string answer4)
        {
            JObject data = new JObject() { { "question", question }, { "correctAns", correctAnswer }, { "ans2", answer2 }, { "ans3", answer3 }, { "ans4", answer4 } };
            return SendRequestDict((byte)RequestCodes.ADD_QUESTION_CODE, data);
        }

        /// <summary>
        /// Will join the hth
        /// {"status" : status}
        /// </summary>
        /// <returns>response</returns>
        public JObject JoinHeadToHead()
        {
            return SendRequestDict((byte)RequestCodes.JOIN_HTH_CODE);
        }

        /// <summary>
        /// Will get hth state
        /// {"status" : status, "hasGameBegun" : hasGameBegun, "questionsAmount" : questionsAmount, "players" : players, "timePerQuestion"}
        /// </summary>
        /// <returns>response</returns>
        public JObject getHeadToHeadState()
        {
            return SendRequestDict((byte)RequestCodes.HTH_GET_STATE_CODE);
        }


        /// <summary>
        /// The function will return a boolean based no if the result given was a success or not.
        /// You can set displayMessageBox to true to see the error in an error message.
        /// In addition you can change customErrorMessage to display an error message of your choice.
        /// </summary>
        /// <param name="response">The response to check</param>
        /// <param name="displayMessageBox">True if you want the error message to display in a message box</param>
        /// <param name="customErrorMessage">a string for a custom error message if needs</param>
        /// <returns>True or False</returns>
        public static bool IsSuccessResponse(JObject response, bool displayMessageBox=true, string customErrorMessage="")
        {
            if ((int)response["code"] == ERROR_CODE)
            {
                if (displayMessageBox)
                {
                    if (customErrorMessage == "")
                    {
                        MessageBox.Show(response["message"]["message"].ToString(), "ERROR", MessageBoxButtons.OK,
                            MessageBoxIcon.Error);
                    }
                    else
                    {
                        MessageBox.Show(customErrorMessage, "ERROR", MessageBoxButtons.OK,
                            MessageBoxIcon.Error);
                    }
                }

                return false;
            }

            return true;
        }
    }
}
